File: src/main/java/com/realting/world/content/combat/strategy/impl/bosses/Brandon.java
package com.realting.world.content.combat.strategy.impl.bosses;
import com.realting.engine.task.Task;
import com.realting.engine.task.TaskManager;
import com.realting.model.Animation;
import com.realting.model.Graphic;
import com.realting.model.Locations;
import com.realting.model.Position;
import com.realting.model.Projectile;
import com.realting.util.Misc;
import com.realting.world.content.combat.CombatContainer;
import com.realting.world.content.combat.CombatType;
import com.realting.world.content.combat.strategy.CombatStrategy;
import com.realting.model.entity.character.CharacterEntity;
import com.realting.model.entity.character.npc.NPC;

public class Brandon implements CombatStrategy {

	public static NPC Brandon;
	private static final Animation attack_anim1 = new Animation(426); 
	private static final Animation attack_anim2 = new Animation(4973);
	private static final Animation attack_anim3 = new Animation(1978);
	private static final Graphic graphic1 = new Graphic(1114);
	private static final Graphic graphic2 = new Graphic(1014);
	private static final Graphic graphic3 = new Graphic(2146);
	private static final Graphic projectile1 = new Graphic(1120);


	public static void spawn() {


		Brandon = new NPC(199, new Position(3023, 3735));
	}

	@Override
	public boolean canAttack(CharacterEntity entity, CharacterEntity victim) {
		return true;
	}

	@Override
	public CombatContainer attack(CharacterEntity entity, CharacterEntity victim) {
		return null;
	}
	
	@Override
	public boolean customContainerAttack(CharacterEntity entity, CharacterEntity victim) {
		NPC Brandon = (NPC)entity;
		if(Brandon.isChargingAttack()) {
			return true;
		}
		int random = Misc.getRandom(10);
		if(random <= 8 && Locations.goodDistance(Brandon.getPosition().getX(), Brandon.getPosition().getY(), victim.getPosition().getX(), victim.getPosition().getY(), 3)) {
			Brandon.performAnimation(attack_anim1);
			Brandon.getCombatBuilder().setContainer(new CombatContainer(Brandon, victim, 1, CombatType.MELEE, true));
			new Projectile(Brandon, victim, projectile1.getId(), 44, 3, 43, 31, 0).sendProjectile();
		} else if(random <= 4 || !Locations.goodDistance(Brandon.getPosition().getX(), Brandon.getPosition().getY(), victim.getPosition().getX(), victim.getPosition().getY(), 8)) {
			Brandon.getCombatBuilder().setContainer(new CombatContainer(Brandon, victim, 1, 3, CombatType.MAGIC, true));
			Brandon.performAnimation(attack_anim3);
			Brandon.setChargingAttack(true);
			TaskManager.submit(new Task(2, Brandon, false) {
				int tick = 0;
				@Override
				public void execute() {
					switch(tick) {
					case 1:
						victim.performGraphic(graphic3);
						Brandon.setChargingAttack(false);
						stop();
						break;
					}
					tick++;
				}
			});
		} else {
			Brandon.getCombatBuilder().setContainer(new CombatContainer(Brandon, victim, 1, CombatType.RANGED, true));
			Brandon.performAnimation(attack_anim2);
			Brandon.setChargingAttack(true);
			TaskManager.submit(new Task(2, Brandon, false) {
				@Override
				public void execute() {
					victim.performGraphic(graphic2);
					Brandon.setChargingAttack(false);
					stop();
				}
			});
		}
		return true;
	}
	
	
	@Override
	public int attackDelay(CharacterEntity entity) {
		return entity.getAttackSpeed();
	}

	@Override
	public int attackDistance(CharacterEntity entity) {
		return 20;
	}

	
	@Override
	public CombatType getCombatType(CharacterEntity entity) {
		return CombatType.MIXED;
	}
}

File: src/main/java/com/realting/world/content/combat/strategy/impl/bosses/Brandon.kt
package com.realting.world.content.combat.strategy.impl.bosses

import com.realting.engine.task.Task
import com.realting.engine.task.TaskManager
import com.realting.model.*
import com.realting.model.entity.character.CharacterEntity
import com.realting.model.entity.character.npc.NPC
import com.realting.util.Misc
import com.realting.world.content.combat.CombatContainer
import com.realting.world.content.combat.CombatType
import com.realting.world.content.combat.strategy.CombatStrategy

class Brandon : CombatStrategy {
    override fun canAttack(entity: CharacterEntity?, victim: CharacterEntity?): Boolean {
        return true
    }

    override fun attack(entity: CharacterEntity?, victim: CharacterEntity?): CombatContainer? {
        return null
    }

    override fun customContainerAttack(entity: CharacterEntity?, victim: CharacterEntity?): Boolean {
        val Brandon = entity as NPC?
        if (Brandon!!.isChargingAttack) {
            return true
        }
        val random = Misc.getRandom(10)
        if (random <= 8 && Locations.goodDistance(
                Brandon.entityPosition.x,
                Brandon.entityPosition.y,
                victim!!.entityPosition.x,
                victim.entityPosition.y,
                3
            )
        ) {
            Brandon.performAnimation(attack_anim1)
            Brandon.combatBuilder.container = CombatContainer(Brandon, victim, 1, CombatType.MELEE, true)
            Projectile(Brandon, victim, projectile1.id, 44, 3, 43, 31, 0).sendProjectile()
        } else if (random <= 4 || !Locations.goodDistance(
                Brandon.entityPosition.x,
                Brandon.entityPosition.y,
                victim!!.entityPosition.x,
                victim.entityPosition.y,
                8
            )
        ) {
            Brandon.combatBuilder.container = CombatContainer(Brandon, victim!!, 1, 3, CombatType.MAGIC, true)
            Brandon.performAnimation(attack_anim3)
            Brandon.isChargingAttack = true
            TaskManager.submit(object : Task(2, Brandon, false) {
                var tick = 0
                public override fun execute() {
                    when (tick) {
                        1 -> {
                            victim.performGraphic(graphic3)
                            Brandon.isChargingAttack = false
                            stop()
                        }
                    }
                    tick++
                }
            })
        } else {
            Brandon.combatBuilder.container = CombatContainer(Brandon, victim, 1, CombatType.RANGED, true)
            Brandon.performAnimation(attack_anim2)
            Brandon.isChargingAttack = true
            TaskManager.submit(object : Task(2, Brandon, false) {
                public override fun execute() {
                    victim.performGraphic(graphic2)
                    Brandon.isChargingAttack = false
                    stop()
                }
            })
        }
        return true
    }

    override fun attackDelay(entity: CharacterEntity?): Int {
        return entity!!.attackSpeed
    }

    override fun attackDistance(entity: CharacterEntity): Int {
        return 20
    }

    override fun getCombatType(entity: CharacterEntity): CombatType? {
        return CombatType.MIXED
    }

    companion object {
        var Brandon: NPC? = null
        private val attack_anim1 = Animation(426)
        private val attack_anim2 = Animation(4973)
        private val attack_anim3 = Animation(1978)
        private val graphic1 = Graphic(1114)
        private val graphic2 = Graphic(1014)
        private val graphic3 = Graphic(2146)
        private val projectile1 = Graphic(1120)
        fun spawn() {
            Brandon = NPC(199, Position(3023, 3735))
        }
    }
}